trio/_util.py

```diff
@@ -138,3 +147,44 @@ def helper(*args, **kwds):
     # A hint for sphinxcontrib-trio:
     helper.__returns_acontextmanager__ = True
     return helper
+
+
+class _UnLockSync:
+    def __init__(self, exc, *args):
+        self._exc = exc
+        self._args = args
+        self._held = False
+
+    def __enter__(self):
+        if self._held:
+            raise self._exc(*self._args)
+        else:
+            self._held = True
+
+    def __exit__(self, *args):
+        self._held = False
+
+
+class UnLock:
```

Comment:
The naming of UnLock (if I understand correctly) seem weird, and counterintuitive. I'm going to guess many people will read it as Unlocking something. Maybe naming as "Restricive" of "Exclusionary" ? Or RaisingLock ?